beacon_node/http_api/src/attestation_rewards.rs

```diff
@@ -0,0 +1,260 @@
+                            effective_balance: *effective_balance_eth as u64,
+                            head: 0,
+                            target: 0,
+                            source: 0,
+                        });
+                match flag_index {
+                    TIMELY_SOURCE_FLAG_INDEX => entry.source += *ideal_reward,
+                    TIMELY_TARGET_FLAG_INDEX => entry.target += *ideal_reward,
+                    TIMELY_HEAD_FLAG_INDEX => entry.head += *ideal_reward,
+                    _ => {}
+                }
+                acc
+            },
+        )
+        .into_values()
+        .collect();
+
+    //--- Calculate total rewards ---//
+    let mut total_rewards = Vec::new();
+
+    let index = participation_cache.eligible_validator_indices();
+
+    for validator_index in index {
+        let eligible = state
+            .is_eligible_validator(previous_epoch, *validator_index)
+            .map_err(|_| {
+                warp_utils::reject::custom_server_error("Unable to get eligible".to_owned())
+            })?;
+
+        let balance = state.get_balance_mut(*validator_index).map_err(|_| {
+            warp_utils::reject::custom_server_error("Unable to get balance".to_owned())
+        })?;
+        *balance = (*balance as f64).round().clamp(0.0, 32.0) as u64;
+
+        let mut head_reward = 0u64;
+        let mut target_reward = 0u64;
+        let mut source_reward = 0u64;
+
+        for &flag_index in [
+            TIMELY_SOURCE_FLAG_INDEX,
+            TIMELY_TARGET_FLAG_INDEX,
+            TIMELY_HEAD_FLAG_INDEX,
+        ]
+        .iter()
+        {
+            if eligible {
+                let voted_correctly = participation_cache
+                    .get_unslashed_participating_indices(flag_index, previous_epoch)
+                    .is_ok();
+                if voted_correctly {
+                    let _ideal_reward = &ideal_rewards
+                        .iter()
+                        .find(|reward| reward.effective_balance == *balance)
+                        .map(|reward| {
+                            head_reward = reward.head;
+                            target_reward = reward.target;
+                            source_reward = reward.source;
+                            reward
+                        })
+                        .unwrap_or(&IdealAttestationRewards {
+                            effective_balance: *balance,
+                            head: 0,
+                            target: 0,
+                            source: 0,
+                        });
+                } else {
+                    match flag_index {
+                        TIMELY_HEAD_FLAG_INDEX => {}
+                        TIMELY_TARGET_FLAG_INDEX => {
+                            target_reward = (-(base_reward as i64 as i128) * weight as i128
+                                / WEIGHT_DENOMINATOR as i128)
+                                as u64
```

Comment:
Rather than casting to `i128` I think we should compute with `u64`s for as long as possible and convert at the last minute

```rust
-1 * base_reward.safe_mul(weight)?.safe_div(WEIGHT_DENOMINATOR)? as i64;
```

We could also share some logic between the cases like:

```rust
let penalty = if flag_index == TIMELY_TARGET_FLAG_INDEX || TIMELY_SOURCE_FLAG_INDEX {
    -1 * base_reward.safe_mul(weight)?.safe_div(WEIGHT_DENOMINATOR)? as i64
} else {
    0
};
if flag_index == TIMELY_TARGET_FLAG_INDEX {
   target_reward = penalty;
} else {
   source_reward = penalty;
}
```